File: 1.21.9/src/main/java/com/bleudev/nine_lifes/util/ComponentUtils.java
package com.bleudev.nine_lifes.util;

import com.bleudev.nine_lifes.custom.CustomTags;
import com.bleudev.nine_lifes.custom.consume.AmethysmConsumeEffect;
import net.minecraft.component.ComponentType;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.*;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Rarity;

import java.util.Objects;
import java.util.function.Predicate;

public class ComponentUtils {
    private static class CheckStackPredicateBuilder {
        private Predicate<ItemStack> predicate;
        private CheckStackPredicateBuilder(Predicate<ItemStack> predicate) {
            this.predicate = predicate;
        }

        public static CheckStackPredicateBuilder of(Predicate<ItemStack> predicate) {
            return new CheckStackPredicateBuilder(predicate);
        }

        public static CheckStackPredicateBuilder create() {
            return of(stack -> true);
        }

        public CheckStackPredicateBuilder of(Item item) {
            return of(this.predicate.and(stack -> stack.isOf(item)));
        }

        public <T> CheckStackPredicateBuilder another_component(ComponentType<T> type, T component) {
            return of(this.predicate.or(stack -> Objects.requireNonNullElse(stack.get(type), "").equals(component)));
        }

        public <T> CheckStackPredicateBuilder or_no_component(ComponentType<T> type) {
            return of(this.predicate.or(stack -> stack.getComponents().contains(type)));
        }

        public CheckStackPredicateBuilder with(Predicate<ItemStack> predicate) {
            return of(this.predicate.and(predicate));
        }

        public CheckStackPredicateBuilder with(CheckStackPredicateBuilder builder) {
            return with(builder.build());
        }

        public Predicate<ItemStack> build() {
            return this.predicate;
        }
    }

    private static final FoodComponent amethyst_shard_food_component = new FoodComponent(3, 0.3f, true);
    private static final ConsumableComponent amethyst_shard_consumable_component = ConsumableComponents.food().consumeEffect(new AmethysmConsumeEffect()).build();
    private static final UseCooldownComponent amethyst_shard_cooldown_component = new UseCooldownComponent(5);
    private static final EnchantableComponent amethyst_enchantable_component = new EnchantableComponent(1);
    private static final Integer amethyst_max_stack_size_component = 65;
    private static final DamageResistantComponent amethyst_damage_resistant_component = new DamageResistantComponent(CustomTags.DamageTypeTags.IS_LIGHTNING_OR_FIRE);

    public static boolean should_update_amethyst_shard(ItemStack stack) {
        return CheckStackPredicateBuilder.create()
            .of(Items.AMETHYST_SHARD)
            .with(CheckStackPredicateBuilder.create()
                .another_component(DataComponentTypes.FOOD, amethyst_shard_food_component)
                .another_component(DataComponentTypes.CONSUMABLE, amethyst_shard_consumable_component)
                .another_component(DataComponentTypes.USE_COOLDOWN, amethyst_shard_cooldown_component)
                .another_component(DataComponentTypes.ENCHANTABLE, amethyst_enchantable_component)
                .another_component(DataComponentTypes.MAX_STACK_SIZE, amethyst_max_stack_size_component)
                .another_component(DataComponentTypes.DAMAGE_RESISTANT, amethyst_damage_resistant_component)
                .or_no_component(DataComponentTypes.ENCHANTMENTS)
            )
        .build().test(stack);
    }

    public static ItemStack item_ensure_custom_foods(ItemStack stack) {
        stack.set(DataComponentTypes.FOOD, amethyst_shard_food_component);
        stack.set(DataComponentTypes.CONSUMABLE, amethyst_shard_consumable_component);
        stack.set(DataComponentTypes.USE_COOLDOWN, amethyst_shard_cooldown_component);
        stack.set(DataComponentTypes.ENCHANTABLE, amethyst_enchantable_component);
        stack.set(DataComponentTypes.MAX_STACK_SIZE, amethyst_max_stack_size_component);
        stack.set(DataComponentTypes.DAMAGE_RESISTANT, amethyst_damage_resistant_component);
        stack.set(DataComponentTypes.RARITY, Rarity.UNCOMMON);
        if (!stack.getComponents().contains(DataComponentTypes.ENCHANTMENTS))
            stack.set(DataComponentTypes.ENCHANTMENTS, ItemEnchantmentsComponent.DEFAULT);
        return stack;
    }

    public static void player_ensure_custom_foods(ServerPlayerEntity player) {
        var inventory = player.getInventory();

        for (int slot = 0; slot < inventory.getMainStacks().size(); slot++) {
            ItemStack stack = inventory.getStack(slot);
            if (should_update_amethyst_shard(stack))
                inventory.setStack(slot, item_ensure_custom_foods(stack));
        }
    }
}

File: 1.21.9/src/main/kotlin/com/bleudev/nine_lifes/util/ComponentUtils.kt
package com.bleudev.nine_lifes.util

import com.bleudev.nine_lifes.custom.NineLifesDamageTypeTags
import com.bleudev.nine_lifes.custom.effect.consume.AmethysmConsumeEffect
import net.minecraft.core.component.DataComponentType
import net.minecraft.core.component.DataComponents
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.food.FoodProperties
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.Rarity
import net.minecraft.world.item.component.Consumables
import net.minecraft.world.item.component.DamageResistant
import net.minecraft.world.item.component.UseCooldown
import net.minecraft.world.item.enchantment.Enchantable
import net.minecraft.world.item.enchantment.ItemEnchantments

private class CheckStackPredicateBuilder private constructor(private val predicate: P<ItemStack>) {
    fun of(item: Item): CheckStackPredicateBuilder = of(this.predicate.and({ stack -> stack.`is`(item) }))
    fun <T> anotherComponent(type: DataComponentType<T>, component: T): CheckStackPredicateBuilder = of(
        this.predicate.or({ stack -> stack.get(type) == component }))
    fun <T> orNoComponent(type: DataComponentType<T>): CheckStackPredicateBuilder = of(
        this.predicate.or({ stack -> stack.getComponents().get(type) != null }))

    fun with(predicate: P<ItemStack>): CheckStackPredicateBuilder = of(this.predicate.and(predicate))
    fun with(builder: CheckStackPredicateBuilder): CheckStackPredicateBuilder = with(builder.build())
    fun build(): P<ItemStack> = this.predicate

    companion object {
        fun of(predicate: P<ItemStack>): CheckStackPredicateBuilder = CheckStackPredicateBuilder(predicate)
        fun create(): CheckStackPredicateBuilder = of(alwaysTrue())
    }
}

private val amethyst_shard_food_component = FoodProperties(3, 0.3f, true)
private val amethyst_shard_consumable_component = Consumables.defaultFood().onConsume(AmethysmConsumeEffect()).build()
private val amethyst_shard_cooldown_component = UseCooldown(5f)
private val amethyst_enchantable_component = Enchantable(1)
private const val amethyst_max_stack_size_component = 65
private val amethyst_damage_resistant_component = DamageResistant(NineLifesDamageTypeTags.IS_LIGHTNING_OR_FIRE)

fun shouldUpdateAmethystShard(stack: ItemStack): Boolean = CheckStackPredicateBuilder.create()
    .of(Items.AMETHYST_SHARD)
    .with(
        CheckStackPredicateBuilder.create()
            .anotherComponent(DataComponents.FOOD, amethyst_shard_food_component)
            .anotherComponent(DataComponents.CONSUMABLE, amethyst_shard_consumable_component)
            .anotherComponent(DataComponents.USE_COOLDOWN, amethyst_shard_cooldown_component)
            .anotherComponent(DataComponents.ENCHANTABLE, amethyst_enchantable_component)
            .anotherComponent(DataComponents.MAX_STACK_SIZE, amethyst_max_stack_size_component)
            .anotherComponent(DataComponents.DAMAGE_RESISTANT, amethyst_damage_resistant_component)
            .orNoComponent(DataComponents.ENCHANTMENTS)
    ).build()(stack)

fun itemEnsureCustomFoods(stack: ItemStack): ItemStack {
    stack.set(DataComponents.FOOD, amethyst_shard_food_component)
    stack.set(DataComponents.CONSUMABLE, amethyst_shard_consumable_component)
    stack.set(DataComponents.USE_COOLDOWN, amethyst_shard_cooldown_component)
    stack.set(DataComponents.ENCHANTABLE, amethyst_enchantable_component)
    stack.set(DataComponents.MAX_STACK_SIZE, amethyst_max_stack_size_component)
    stack.set(DataComponents.DAMAGE_RESISTANT, amethyst_damage_resistant_component)
    stack.set(DataComponents.RARITY, Rarity.UNCOMMON)
    if (stack.get(DataComponents.ENCHANTMENTS) == null) stack.set(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY)
    return stack
}

fun playerEnsureCustomFoods(player: ServerPlayer) {
    val inventory = player.getInventory()
    for (slot in 0..<inventory.nonEquipmentItems.size) {
        val stack = inventory.getItem(slot)
        if (shouldUpdateAmethystShard(stack)) inventory.setItem(slot, itemEnsureCustomFoods(stack))
    }
}